src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -318,6 +306,7 @@ private void EventLoop()
             try
             {
                 bool shutdown = false;
+                SocketAsyncContext[] contexts = new SocketAsyncContext[EventBufferCount];
```

Comment:
Could we make _handleToContextMap a `ConcurrentDictionary<IntPtr, SocketAsyncContext>`?  That would result in an allocation on every add, but adds are rare-ish, only when new sockets are added (right?), in which case there's already other allocation happening (e.g. the socket itself and all associated state), and we could avoid the lock on reads entirely.

Reply:
I'll make a branch that implements this and we can benchmark both.

Reply:
This branch uses `ConcurrentDictionary`: https://github.com/tmds/corefx/commit/9cc752708ff0617f9fc257b2aeaf71face32b3e0. @stephentoub , you can add some review comments on that commit if you want.

Reply:
Thanks.  I prefer the ConcurrentDictionary version, but we should see what perf looks like for both.